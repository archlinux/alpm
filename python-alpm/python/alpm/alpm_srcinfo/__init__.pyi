"""A module for parsing and linting of ALPM SRCINFO files."""

from . import error, source_info
from .error import SourceInfoError
from .source_info.v1 import SourceInfoV1
from .source_info.v1.merged import MergedPackage

__all__ = [
    "SourceInfoError",
    "error",
    "source_info",
    "SourceInfoV1",
    "MergedPackage",
]
