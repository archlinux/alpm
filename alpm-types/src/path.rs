use std::{
    fmt::{Display, Formatter},
    path::{Path, PathBuf},
    str::FromStr,
};

use serde::{Deserialize, Serialize};

use crate::Error;

/// A representation of an absolute path
///
/// AbsolutePath wraps a `PathBuf`, that is guaranteed to be absolute.
///
/// ## Examples
/// ```
/// use std::{path::PathBuf, str::FromStr};
///
/// use alpm_types::{AbsolutePath, Error};
///
/// # fn main() -> Result<(), alpm_types::Error> {
/// // Create AbsolutePath from &str
/// assert_eq!(
///     AbsolutePath::from_str("/"),
///     AbsolutePath::new(PathBuf::from("/"))
/// );
/// assert_eq!(
///     AbsolutePath::from_str("./"),
///     Err(Error::PathNotAbsolute(PathBuf::from("./")))
/// );
///
/// // Format as String
/// assert_eq!("/", format!("{}", AbsolutePath::from_str("/")?));
/// # Ok(())
/// # }
/// ```
#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct AbsolutePath(PathBuf);

impl AbsolutePath {
    /// Create a new `AbsolutePath`
    pub fn new(path: PathBuf) -> Result<AbsolutePath, Error> {
        match path.is_absolute() {
            true => Ok(AbsolutePath(path)),
            false => Err(Error::PathNotAbsolute(path)),
        }
    }

    /// Return a reference to the inner type
    pub fn inner(&self) -> &Path {
        &self.0
    }
}

impl FromStr for AbsolutePath {
    type Err = Error;

    /// Parses an absolute path from a string
    ///
    /// # Errors
    ///
    /// Returns an error if the path is not absolute
    fn from_str(s: &str) -> Result<AbsolutePath, Self::Err> {
        match Path::new(s).is_absolute() {
            true => Ok(AbsolutePath(PathBuf::from(s))),
            false => Err(Error::PathNotAbsolute(PathBuf::from(s))),
        }
    }
}

impl Display for AbsolutePath {
    fn fmt(&self, fmt: &mut Formatter) -> std::fmt::Result {
        write!(fmt, "{}", self.inner().display())
    }
}

/// An absolute path used as build directory
///
/// This is a type alias for [`AbsolutePath`]
///
/// ## Examples
/// ```
/// use std::str::FromStr;
///
/// use alpm_types::{Error, BuildDirectory};
///
/// # fn main() -> Result<(), alpm_types::Error> {
/// // Create BuildDirectory from &str and format it
/// assert_eq!(
///     "/etc",
///     BuildDirectory::from_str("/etc")?.to_string()
/// );
/// # Ok(())
/// # }
pub type BuildDirectory = AbsolutePath;

/// An absolute path used as start directory in a package build environment
///
/// This is a type alias for [`AbsolutePath`]
///
/// ## Examples
/// ```
/// use std::str::FromStr;
///
/// use alpm_types::{Error, StartDirectory};
///
/// # fn main() -> Result<(), alpm_types::Error> {
/// // Create StartDirectory from &str and format it
/// assert_eq!(
///     "/etc",
///     StartDirectory::from_str("/etc")?.to_string()
/// );
/// # Ok(())
/// # }
pub type StartDirectory = AbsolutePath;

/// A representation of a relative file path
///
/// `RelativePath` wraps a `PathBuf` that is guaranteed to represent a
/// relative file path (i.e. it does not end with a `/`).
///
/// ## Examples
///
/// ```
/// use std::{path::PathBuf, str::FromStr};
///
/// use alpm_types::{Error, RelativePath};
///
/// # fn main() -> Result<(), alpm_types::Error> {
/// // Create RelativePath from &str
/// assert_eq!(
///     RelativePath::from_str("etc/test.conf"),
///     RelativePath::new(PathBuf::from("etc/test.conf"))
/// );
/// assert_eq!(
///     RelativePath::from_str("/etc/test.conf"),
///     Err(Error::PathNotRelative(PathBuf::from("/etc/test.conf")))
/// );
///
/// // Format as String
/// assert_eq!(
///     "test/test.txt",
///     RelativePath::from_str("test/test.txt")?.to_string()
/// );
/// # Ok(())
/// # }
/// ```
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct RelativePath(PathBuf);

impl RelativePath {
    /// Create a new `RelativePath`
    pub fn new(path: PathBuf) -> Result<RelativePath, Error> {
        match path.is_relative()
            && !path
                .to_string_lossy()
                .ends_with(std::path::MAIN_SEPARATOR_STR)
        {
            true => Ok(RelativePath(path)),
            false => Err(Error::PathNotRelative(path)),
        }
    }

    /// Return a reference to the inner type
    pub fn inner(&self) -> &Path {
        &self.0
    }
}

impl FromStr for RelativePath {
    type Err = Error;

    /// Parses a relative path from a string
    ///
    /// # Errors
    ///
    /// Returns an error if the path is not relative
    fn from_str(s: &str) -> Result<RelativePath, Self::Err> {
        Self::new(PathBuf::from(s))
    }
}

impl Display for RelativePath {
    fn fmt(&self, fmt: &mut Formatter) -> std::fmt::Result {
        write!(fmt, "{}", self.inner().display())
    }
}

/// The path of a packaged file that should be preserved during package operations
///
/// This is a type alias for [`RelativePath`]
///
/// ## Examples
/// ```
/// use std::str::FromStr;
///
/// use alpm_types::Backup;
///
/// # fn main() -> Result<(), alpm_types::Error> {
/// // Create Backup from &str and format it
/// assert_eq!(
///     "etc/test.conf",
///     Backup::from_str("etc/test.conf")?.to_string()
/// );
/// # Ok(())
/// # }
pub type Backup = RelativePath;

/// A special install script that is to be included in the package
///
/// This is a type alias for [RelativePath`]
///
/// ## Examples
/// ```
/// use std::str::FromStr;
///
/// use alpm_types::{Error, Install};
///
/// # fn main() -> Result<(), alpm_types::Error> {
/// // Create Install from &str and format it
/// assert_eq!(
///     "scripts/setup.install",
///     Install::from_str("scripts/setup.install")?.to_string()
/// );
/// # Ok(())
/// # }
pub type Install = RelativePath;

/// The relative path to a changelog file that may be included in a package
///
/// This is a type alias for [`RelativePath`]
///
/// ## Examples
/// ```
/// use std::str::FromStr;
///
/// use alpm_types::{Error, Changelog};
///
/// # fn main() -> Result<(), alpm_types::Error> {
/// // Create Changelog from &str and format it
/// assert_eq!(
///     "changelog.md",
///     Changelog::from_str("changelog.md")?.to_string()
/// );
/// # Ok(())
/// # }
pub type Changelog = RelativePath;

#[cfg(test)]
mod tests {
    use rstest::rstest;

    use super::*;

    #[rstest]
    #[case("/home", BuildDirectory::new(PathBuf::from("/home")))]
    #[case("./", Err(Error::PathNotAbsolute(PathBuf::from("./"))))]
    #[case("~/", Err(Error::PathNotAbsolute(PathBuf::from("~/"))))]
    #[case("foo.txt", Err(Error::PathNotAbsolute(PathBuf::from("foo.txt"))))]
    fn build_dir_from_string(#[case] s: &str, #[case] result: Result<BuildDirectory, Error>) {
        assert_eq!(BuildDirectory::from_str(s), result);
    }

    #[rstest]
    #[case("/start", StartDirectory::new(PathBuf::from("/start")))]
    #[case("./", Err(Error::PathNotAbsolute(PathBuf::from("./"))))]
    #[case("~/", Err(Error::PathNotAbsolute(PathBuf::from("~/"))))]
    #[case("foo.txt", Err(Error::PathNotAbsolute(PathBuf::from("foo.txt"))))]
    fn startdir_from_str(#[case] s: &str, #[case] result: Result<StartDirectory, Error>) {
        assert_eq!(StartDirectory::from_str(s), result);
    }

    #[rstest]
    #[case("etc/test.conf", RelativePath::new(PathBuf::from("etc/test.conf")))]
    #[case(
        "/etc/test.conf",
        Err(Error::PathNotRelative(PathBuf::from("/etc/test.conf")))
    )]
    #[case("etc/", Err(Error::PathNotRelative(PathBuf::from("etc/"))))]
    #[case("etc", RelativePath::new(PathBuf::from("etc")))]
    #[case(
        "../etc/test.conf",
        RelativePath::new(PathBuf::from("../etc/test.conf"))
    )]
    fn relative_path_from_str(#[case] s: &str, #[case] result: Result<RelativePath, Error>) {
        assert_eq!(RelativePath::from_str(s), result);
    }
}
