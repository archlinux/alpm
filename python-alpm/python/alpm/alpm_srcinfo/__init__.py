# TODO: <https://gitlab.archlinux.org/archlinux/alpm/alpm/-/issues/204>
from alpm import alpm_srcinfo as _mod

__all__ = _mod.__all__
for name in getattr(_mod, "__all__"):
    globals()[name] = getattr(_mod, name)
