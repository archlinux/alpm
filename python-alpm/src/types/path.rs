use std::path::PathBuf;

use pyo3::prelude::*;

#[pyclass(frozen, eq)]
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct RelativePath(alpm_types::RelativePath);

#[pymethods]
impl RelativePath {
    // PyO3 handles the conversion from Rust's `std::path::PathBuf` to Python's `pathlib.Path`.
    #[new]
    fn new(path: PathBuf) -> Result<Self, crate::types::Error> {
        let inner = alpm_types::RelativePath::new(path)?;
        Ok(Self(inner))
    }

    fn __str__(&self) -> String {
        self.0.to_string()
    }

    fn __repr__(&self) -> String {
        format!("RelativePath('{}')", self.0)
    }
}
