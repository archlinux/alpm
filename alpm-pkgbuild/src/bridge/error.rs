//! The error types used in the scope of `alpm-pkgbuild-bridge` output logic.

#[cfg(doc)]
use alpm_srcinfo::SourceInfo;
use alpm_types::{Architecture, Name};
use thiserror::Error;
use winnow::error::{ContextError, ParseError};

use super::parser::Keyword;

/// A lower-level error that may occur when converting `alpm-pkgbuild-bridge` script output into the
/// [`SourceInfo`] format.
#[derive(Debug, Error)]
pub enum BridgeError {
    /// No `pkgname` has been specified.
    #[error("No 'pkgname' has been specified. At least one must be given.")]
    NoName,

    /// A package name is not valid.
    #[error("The package name '{name}' is not valid:\n{error}")]
    InvalidPackageName {
        /// The invalid package name.
        name: String,
        /// The source error.
        error: alpm_types::Error,
    },

    /// A `package` function has been declared for a split package, but it is not defined in
    /// `pkgname`.
    #[error(
        "The split package '{0}' is not declared in pkgname, but a package function is present for it."
    )]
    UndeclaredPackageName(String),

    /// An unused package function exists for an undeclared [alpm-split-package].
    ///
    /// [alpm-split-package]: https://alpm.archlinux.page/specifications/alpm-split-package.7.html
    #[error("An unused package function exists for undeclared split package: '{0}'")]
    UnusedPackageFunction(Name),

    /// A type parser fails on a certain keyword.
    #[error("Missing keyword: '{keyword}'")]
    MissingRequiredKeyword {
        /// The keyword that cannot be parsed.
        keyword: Keyword,
    },

    /// A type parser fails on a certain keyword.
    #[error("Failed to parse input for keyword '{keyword}':\n{error}")]
    ParseError {
        /// The keyword that cannot be parsed.
        keyword: Keyword,
        /// The error message.
        error: String,
    },

    /// A variable is expected to be of a different type.
    /// E.g. `String` when an `Array` is expected.
    #[error(
        "Got wrong variable type for keyword '{keyword}'. Expected a {expected}, got a {actual}"
    )]
    WrongVariableType {
        /// The name of the keyword for which a wrong variable type is used.
        keyword: String,
        /// The expected type of variable.
        expected: String,
        /// The actual type of variable.
        actual: String,
    },

    /// A keyword has an architecture suffix even though it shouldn't have one.
    #[error("Found unexpected architecture suffix '{suffix}' for keyword '{keyword}'")]
    UnexpectedArchitecture {
        /// The keyword for which an unexpected architecture suffix is found.
        keyword: Keyword,
        /// The architecture that is found for the `keyword`.
        suffix: Architecture,
    },

    /// A keyword that cannot be cleared is attempted to be cleared.
    #[error("Tried to clear value for keyword '{keyword}', which is not allowed.")]
    UnclearableValue {
        /// The keyword that is attempted to be cleared.
        keyword: Keyword,
    },

    /// A keyword should have only a single value, but an array is found.
    #[error(
        "Found array of values for keyword '{keyword}' that expects a single value:\n{}",
        values.iter().map(|s| format!("\"{s}\"")).collect::<Vec<String>>().join(", ")
    )]
    UnexpectedArray {
        /// The keyword for which a single value should be used.
        keyword: Keyword,
        /// The values that are used for the `keyword`.
        values: Vec<String>,
    },

    /// A duplicate [`Architecture`] is specified.
    #[error("Found duplicate architecture: {duplicate}")]
    DuplicateArchitecture {
        /// The duplicate architecture.
        duplicate: Architecture,
    },
}

impl<'a> From<(Keyword, ParseError<&'a str, ContextError>)> for BridgeError {
    /// Converts a tuple of ([`Keyword`] and [`ParseError`]) into a [`BridgeError::ParseError`].
    fn from(value: (Keyword, ParseError<&'a str, ContextError>)) -> Self {
        Self::ParseError {
            keyword: value.0,
            error: value.1.to_string(),
        }
    }
}
