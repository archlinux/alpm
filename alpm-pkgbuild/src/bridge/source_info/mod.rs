//! Convert untyped and unchecked [`BridgeOutput`] into a well-formed [`SourceInfoV1`].

mod package;
mod package_base;

use std::collections::HashMap;

use alpm_srcinfo::SourceInfoV1;
use alpm_types::{Architecture, Name};
use package::handle_packages;
use package_base::handle_package_base;
use winnow::{
    Parser,
    error::{ContextError, ErrMode, ParseError},
};

use super::{
    error::BridgeError,
    parser::{BridgeOutput, Keyword, Value},
};

impl TryFrom<BridgeOutput> for SourceInfoV1 {
    type Error = BridgeError;

    /// Creates a [`SourceInfoV1`] from a [`BridgeOutput`].
    ///
    /// # Errors
    ///
    /// Returns an error if
    ///
    /// - a required field is not set,
    /// - a `package` functions exists, but does not correspond to a declared [alpm-split-package],
    /// - a `package` function without an [alpm-package-name] suffix exists in an
    ///   [alpm-split-package] setup,
    /// - a value cannot be turned into its [`alpm_types`] equivalent,
    /// - multiple values exist for a field that only accepts a singular value,
    /// - an [alpm-architecture] is duplicated,
    /// - an [alpm-architecture] is cleared in `package` function,
    /// - or an [alpm-architecture] suffix is set on a keyword that does not support it.
    ///
    /// [alpm-architecture]: https://alpm.archlinux.page/specifications/alpm-architecture.7.html
    /// [alpm-package-name]: https://alpm.archlinux.page/specifications/alpm-package-name.7.html
    /// [alpm-split-package]: https://alpm.archlinux.page/specifications/alpm-split-package.7.html
    fn try_from(mut value: BridgeOutput) -> Result<Self, Self::Error> {
        let mut name = None;
        // Check if there's a `pkgbase` section, which hints that this is a split package.
        let pkgbase_keyword = Keyword::simple("pkgbase");
        if let Some(value) = value.package_base.remove(&pkgbase_keyword) {
            name = Some(parse_value(&pkgbase_keyword, &value, Name::parser)?);
        }

        // Get the list of all packages that are declared.
        let pkgname_keyword = Keyword::simple("pkgname");
        let names = ensure_keyword_exists(&pkgname_keyword, &mut value.package_base)?;
        let names = parse_value_array(&pkgname_keyword, &names, Name::parser)?;

        // Use the `pkgbase` name by default, otherwise fallback to the first `pkgname` entry.
        let name = match name {
            Some(name) => name,
            None => {
                // The first package name is used as the name for the pkgbase section.
                names.first().cloned().ok_or(BridgeError::NoName)?
            }
        };

        let base = handle_package_base(name.clone(), value.package_base)?;

        // Go through all declared functions and ensure that the package functions are also
        // declared via `pkgname`. If one of them is not, this is a bug.
        for name in value.functions {
            let Some(name) = name.0 else { continue };

            let name =
                Name::parser
                    .parse(&name)
                    .map_err(|err| BridgeError::InvalidPackageName {
                        name: name.clone(),
                        error: err.into(),
                    })?;

            if !names.contains(&name) {
                return Err(BridgeError::UndeclaredPackageName(name.to_string()));
            }
        }

        let packages = handle_packages(name, names, value.packages)?;

        Ok(SourceInfoV1 { base, packages })
    }
}

/// Ensures a [`Keyword`] exists in a [`HashMap`], removes it and returns it.
///
/// This is a helper function to ensure expected values are set while throwing context-rich
/// errors if they don't.
///
/// # Errors
///
/// Returns an error if `keyword` is not a key in `map`.
fn ensure_keyword_exists(
    keyword: &Keyword,
    map: &mut HashMap<Keyword, Value>,
) -> Result<Value, BridgeError> {
    match map.remove(keyword) {
        Some(value) => Ok(value),
        None => Err(BridgeError::MissingRequiredKeyword {
            keyword: keyword.clone(),
        }),
    }
}

/// Ensures that a combination of a [`Keyword`] and an optional [`Architecture`] does not use an
/// [`Architecture`].
///
/// # Errors
///
/// Returns an error, if `architecture` provides an [`Architecture`].
fn ensure_no_suffix(
    keyword: &Keyword,
    architecture: Option<Architecture>,
) -> Result<(), BridgeError> {
    if let Some(arch) = architecture {
        return Err(BridgeError::UnexpectedArchitecture {
            keyword: keyword.clone(),
            suffix: arch,
        });
    }

    Ok(())
}

/// Ensures that a combination of [`Keyword`] and [`Value`] uses a [`Value::Single`] and returns the
/// value.
///
/// # Errors
///
/// Returns an error, if `value` is a [`Value::Array`].
fn ensure_single_value<'a>(keyword: &Keyword, value: &'a Value) -> Result<&'a String, BridgeError> {
    match value {
        Value::Single(item) => Ok(item),
        Value::Array(values) => Err(BridgeError::UnexpectedArray {
            keyword: keyword.clone(),
            values: values.clone(),
        }),
    }
}

/// Ensures that a combination of [`Keyword`] and [`Value`] uses a [`Value::Single`] and parses the
/// value as a specific type.
///
/// # Errors
///
/// Returns a error for `keyword` if `value` is not [`Value::Single`] or cannot be parsed as the
/// specific type.
fn parse_value<'a, O, P: Parser<&'a str, O, ErrMode<ContextError>>>(
    keyword: &Keyword,
    value: &'a Value,
    mut parser: P,
) -> Result<O, BridgeError> {
    let input = ensure_single_value(keyword, value)?;
    Ok(parser.parse(input).map_err(|err| (keyword.clone(), err))?)
}

/// Ensures a combination of [`Keyword`] and [`Value`] uses a [`Value::Single`] and parses the value
/// as a specific, but optional type.
///
/// Returns [`None`] if `value` is **empty**.
///
/// # Errors
///
/// Returns a error for `keyword` if `value` is not [`Value::Single`] or cannot be parsed as the
/// specific type.
fn parse_optional_value<'a, O, P: Parser<&'a str, O, ErrMode<ContextError>>>(
    keyword: &Keyword,
    value: &'a Value,
    mut parser: P,
) -> Result<Option<O>, BridgeError> {
    let input = ensure_single_value(keyword, value)?;

    if input.trim().is_empty() {
        return Ok(None);
    }

    Ok(Some(
        parser.parse(input).map_err(|err| (keyword.clone(), err))?,
    ))
}

/// Parses a [`Value`] as a [`Vec`] of specific types.
///
/// Does not differentiate between [`Value::Single`] and [`Value::Array`] variants, as a
/// [`PKGBUILD`] allows either for array values.
///
/// # Errors
///
/// Returns a error for `keyword` if `value` cannot be parsed.
///
/// [`PKGBUILD`]: https://man.archlinux.org/man/PKGBUILD.5
fn parse_value_array<'a, O, P: Parser<&'a str, O, ErrMode<ContextError>>>(
    keyword: &Keyword,
    value: &'a Value,
    mut parser: P,
) -> Result<Vec<O>, BridgeError> {
    let input = value.as_vec();
    Ok(input
        .into_iter()
        .map(|item| parser.parse(item).map_err(|err| (keyword.clone(), err)))
        .collect::<Result<Vec<O>, (Keyword, ParseError<&'a str, ContextError>)>>()?)
}
