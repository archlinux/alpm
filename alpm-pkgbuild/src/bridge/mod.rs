//! The `PKGBUILD` to `.SRCINFO` bridge logic.

pub mod error;
pub(crate) mod parser;
pub mod source_info;

use std::{
    io::ErrorKind,
    path::Path,
    process::{Command, Stdio},
};

use log::debug;
use which::which;

use crate::error::Error;

const DEFAULT_SCRIPT_NAME: &str = "alpm-pkgbuild-bridge";

/// Runs the [`alpm-pkgbuild-bridge`] script, which exposes all relevant information of a
/// [`PKGBUILD`] in a custom format.
///
/// Returns the output of that script as a `String`.
///
/// # Examples
///
/// ```
/// use std::{fs::File, io::Write};
///
/// use alpm_pkgbuild::bridge::run_bridge_script;
/// use tempfile::tempdir;
///
/// const TEST_FILE: &str = include_str!("../../tests/test_files/normal.pkgbuild");
///
/// # fn main() -> testresult::TestResult {
/// // Create a temporary directory where we write the PKGBUILD to.
/// let temp_dir = tempdir()?;
/// let path = temp_dir.path().join("PKGBUILD");
/// let mut file = File::create(&path)?;
/// file.write_all(TEST_FILE.as_bytes());
///
/// // Call the bridge script on that path.
/// println!("{}", run_bridge_script(&path)?);
/// # Ok(())
/// # }
/// ```
///
/// # Errors
///
/// Returns an error if
///
/// - `pkgbuild_path` does not exist,
/// - `pkgbuild_path` does not have a file name,
/// - `pkgbuild_path` is not a file,
/// - or running the `alpm-pkgbuild-bridge` script fails.
///
/// [`PKGBUILD`]: https://man.archlinux.org/man/PKGBUILD.5
/// [`alpm-pkgbuild-bridge`]: https://gitlab.archlinux.org/archlinux/alpm/alpm-pkgbuild-bridge
pub fn run_bridge_script(pkgbuild_path: &Path) -> Result<String, Error> {
    // Make sure the PKGBUILD path exists.
    if !pkgbuild_path.exists() {
        let source = std::io::Error::new(ErrorKind::NotFound, "No such file or directory.");
        return Err(Error::IoPath {
            path: pkgbuild_path.to_path_buf(),
            context: "checking for PKGBUILD",
            source,
        });
    }

    // Make sure the PKGBUILD path contains a filename.
    let Some(filename) = pkgbuild_path.file_name() else {
        return Err(Error::InvalidFile {
            path: pkgbuild_path.to_owned(),
            context: "No filename provided in path",
        });
    };

    // Make sure the PKGBUILD path actually points to a file.
    let metadata = pkgbuild_path.metadata().map_err(|source| Error::IoPath {
        path: pkgbuild_path.to_owned(),
        context: "getting metadata of file",
        source,
    })?;
    if !metadata.file_type().is_file() {
        return Err(Error::InvalidFile {
            path: pkgbuild_path.to_owned(),
            context: "Path doesn't point to a file.",
        });
    };

    let script_path = which(DEFAULT_SCRIPT_NAME).map_err(|source| Error::ScriptNotFound {
        script_name: DEFAULT_SCRIPT_NAME.to_string(),
        source,
    })?;

    let mut command = Command::new(script_path);
    // Change the CWD to the directory that contains the PKGBUILD
    if let Some(parent) = pkgbuild_path.parent() {
        // `parent` returns an empty path for relative paths with a single component.
        if parent != Path::new("") {
            command.current_dir(parent);
        }
    }

    let parameters = vec![filename.to_string_lossy().to_string()];
    command.args(&parameters);

    command.stdout(Stdio::piped());
    command.stderr(Stdio::piped());

    debug!(
        "Spawning command '{DEFAULT_SCRIPT_NAME} {}'",
        parameters.join(" ")
    );
    let child = command.spawn().map_err(|source| Error::ScriptError {
        context: "spawn",
        parameters: parameters.clone(),
        source,
    })?;

    debug!("Waiting for '{DEFAULT_SCRIPT_NAME}' to finish");
    let output = child
        .wait_with_output()
        .map_err(|source| Error::ScriptError {
            context: "finish",
            parameters: parameters.clone(),
            source,
        })?;

    if !output.status.success() {
        let stdout = String::from_utf8_lossy(&output.stdout).to_string();
        let stderr = String::from_utf8_lossy(&output.stderr).to_string();
        return Err(Error::ScriptExecutionError {
            parameters,
            stdout,
            stderr,
        });
    }

    String::from_utf8(output.stdout).map_err(Error::from)
}
