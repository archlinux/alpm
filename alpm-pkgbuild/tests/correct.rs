//! Happy-path tests for the PKGBUILD to SRCINFO conversion.

// NOTE: With rstest > 0.25.0 this can be removed!
#![allow(missing_docs)]

use std::path::PathBuf;

use alpm_pkgbuild::bridge::{BridgeOutput, run_bridge_script};
use alpm_srcinfo::SourceInfoV1;
use insta::assert_snapshot;
use rstest::rstest;
use testresult::TestResult;

/// Get some valid PKGBUILD files and make sure the generated SRCINFO output is correct.
///
/// This tests the whole pipeline by also inspecting and snapshotting the intermediate output
/// from the `alpm-pkgbuild-bridge`
#[rstest]
pub fn correct_files(#[files("tests/correct/*.pkgbuild")] case: PathBuf) -> TestResult {
    let test_name = case.file_stem().unwrap().to_str().unwrap().to_string();

    // Run the bridge script on the input file.
    let raw_bridge_output = run_bridge_script(&case)?;

    // Make sure the generated bridge output matches the expected values.
    insta::with_settings!({
        description => format!("{test_name} PKGBUILD -> SRCINFO generation."),
        snapshot_path => "correct_bridge_output",
        prepend_module_to_snapshot => false,
    }, {
        assert_snapshot!(format!("{test_name}_bridge"), raw_bridge_output);
    });

    // Take the raw bridge script output and parse it.
    let output = BridgeOutput::from_script_output(&raw_bridge_output)?;
    // Then convert it into a SourceInfo struct.
    let source_info: SourceInfoV1 = output.try_into()?;

    // Now create actual .SRCINFO file format output.
    let srcinfo_output = source_info.as_srcinfo();

    // Compare the generated source_info json with the expected snapshot.
    // Remove the usual module prefix by explicitly setting the snapshot path.
    // This is necessary, as we're manually sorting snapshots by test scenario.
    insta::with_settings!({
        description => format!("{test_name} PKGBUILD -> SRCINFO generation."),
        snapshot_path => "correct_snapshots",
        prepend_module_to_snapshot => false,
    }, {
        assert_snapshot!(format!("{test_name}_srcinfo"), srcinfo_output  );
    });

    Ok(())
}
