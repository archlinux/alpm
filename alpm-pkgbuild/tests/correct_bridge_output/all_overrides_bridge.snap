---
source: alpm-pkgbuild/tests/correct.rs
description: all_overrides PKGBUILD -> SRCINFO generation.
expression: raw_bridge_output
---
VAR GLOBAL ARRAY arch "x86_64" "aarch64"
VAR GLOBAL ARRAY backup "etc/pacman.conf"
VAR GLOBAL STRING changelog "changelog.stub"
VAR GLOBAL ARRAY groups "group" "group_2"
VAR GLOBAL STRING epoch "1"
VAR GLOBAL STRING install "install.sh.stub"
VAR GLOBAL ARRAY license "MIT"
VAR GLOBAL ARRAY options "!lto"
VAR GLOBAL STRING pkgbase "example"
VAR GLOBAL STRING pkgdesc "A example with all pkgbase properties set."
VAR GLOBAL ARRAY pkgname "example"
VAR GLOBAL STRING pkgrel "1"
VAR GLOBAL STRING pkgver "0.1.0"
VAR GLOBAL STRING url "https://archlinux.org/"
VAR GLOBAL ARRAY conflicts "default_conflict"
VAR GLOBAL ARRAY optdepends "default_optdep"
VAR GLOBAL ARRAY provides "default_provides"
VAR GLOBAL ARRAY replaces "default_replaces"
VAR GLOBAL ARRAY conflicts_x86_64 "arch_default_conflict"
VAR GLOBAL ARRAY depends_x86_64 "arch_default_dep"
VAR GLOBAL ARRAY optdepends_x86_64 "arch_default_optdep"
VAR GLOBAL ARRAY provides_x86_64 "arch_default_provides"
VAR GLOBAL ARRAY replaces_x86_64 "arch_default_replaces"
VAR FUNCTION package_example ARRAY arch "aarch64"
VAR FUNCTION package_example ARRAY backup "overridden"
VAR FUNCTION package_example STRING changelog "overridden.stub"
VAR FUNCTION package_example ARRAY groups "overridden"
VAR FUNCTION package_example STRING install "overridden.stub"
VAR FUNCTION package_example ARRAY license "Apache-2.0"
VAR FUNCTION package_example ARRAY options "emptydirs"
VAR FUNCTION package_example STRING pkgdesc "overridden"
VAR FUNCTION package_example STRING url "https://overridden.com/"
VAR FUNCTION package_example ARRAY conflicts "overridden"
VAR FUNCTION package_example STRING depends ""
VAR FUNCTION package_example ARRAY optdepends
VAR FUNCTION package_example ARRAY provides "overridden"
VAR FUNCTION package_example ARRAY replaces "overridden"
VAR FUNCTION package_example ARRAY conflicts_x86_64 "arch_overridden"
VAR FUNCTION package_example ARRAY depends_x86_64 "arch_overridden"
VAR FUNCTION package_example ARRAY optdepends_x86_64 "arch_overridden"
VAR FUNCTION package_example ARRAY provides_x86_64 "arch_overridden"
VAR FUNCTION package_example ARRAY replaces_x86_64 "arch_overridden"
FUNCTION package_example
